### ISSUE TITLE
Don't use extent for selecting icons (was: Save updated extents of icons in diagrams when resizing icon extents)

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#11866]
@adeas31, please consider the following MWE
```modelica
package TestChangeIconExtent
    model M
  equation

  annotation(
      Icon(graphics = {Rectangle(fillPattern = FillPattern.Solid, extent = {{-80, 20}, {80, -20}}), Rectangle(extent = {{-54, 20}, {-54, 20}})}));
end M;

  model S
    M m1 annotation(
      Placement(transformation(extent = {{-10, -10}, {10, 10}})));
    M m2 annotation(
      Placement(transformation(origin = {0, 40}, extent = {{-10, -10}, {10, 10}}, rotation = 90)));
  end S;
end TestChangeIconExtent;
```
Open the model `S`: the two icons of `m1` and `m2` are shown.
![immagine](https://github.com/OpenModelica/OpenModelica/assets/4233148/a6ee1cd4-ee4a-4611-9d92-ad2ab91690d4)

Now open the model `M`, switch to the icon layer and change the bottom and top coordinates of the icon's extent to 20 and -20. If you now switch back to the diagram view of `S`, it still looks fine. I guess that is because the view was cached and not updated. However, if you save the package, unload it, and reload it, it will look like this:
![immagine](https://github.com/OpenModelica/OpenModelica/assets/4233148/41b76e6a-7c62-4b94-8696-bc42f541df21)
This is because the extent of the icons of `m1` and `m2` is unchanged, but now applies to an icon which has a squat extent instead of a square one, so they are stretched in the vertical direction.

Would it be possible to update the extent of icons in diagrams of loaded models when you change the extent of an icon, so that the final appearance remains the same?

The use case here is PowerGrids: we realized that the extent of buses is too large and should be reduced in the vertical direction; however, if we do so, all the buses in all the examples of the library get stretched in the vertical direction. We can fix this manually, but it's a horribly tedious work.

Thanks!

**Comment by adeas31** [comment]
> Would it be possible to update the extent of icons in diagrams of loaded models when you change the extent of an icon, so that the final appearance remains the same?

The icons of `m1` and `m2` should be changed accordingly. So I guess some update call is missing here. But to actually change the extent of `m1` and `m2` in the Modelica code is not possible. I don't think any Modelica tool does that.

**Comment by adeas31** [comment]
#11872 only fixes the visual issue i.e., now the components will be updated visually when the icon coordinate system extents are changed.

**Comment by casella** [comment]
Two steps are needed:
1. the frontend should identify which annotations must be updated
2. OMEdit should compute the new extent

**Comment by casella** [comment]
For step 1., we need something along the lines of this algorithm in pseudo-code:
- given a certain class C with Icon annotation (for which we are changing the extent)
- for each loaded class
  - for each component that has a Diagram annotation
    - update := false
    - if the component instantiates C then update := true
    - if the component instantiates a subclass of C then  
      - update := true
      - for each subclass C_s from top to bottom
        - if C_s has an Icon(extent) annotation, then update := false

This is because one may be changing the extent in a base class, but if a child class has another extent annotation, I understand it overrides the one of the ancestor(s), so we shouldn't bother.

@perost how do you see that? Is it doable?

**Comment by casella** [comment]
For step 2., I'm working on it

**Comment by casella** [comment]
@adeas31, Step 2. is actually almost trivial. You should already have all the code in place that takes the coordinates of graphical primitives (lines, circles, text, whatever) in the icon reference frame and transforms them into the diagram reference frame, where you actually draw them. So, what you need to do is:

- Use it to compute the coordinates where a point located in the bottom left corner of the updated Icon extent will be plotted in the diagram
- Use it to compute the coordinates where a point located in the upper left corner of the updated Icon extent will be plotted in the diagram
- Update the extent of the icon in the diagram so that it matches those coordinates

That's it. Doing this, it is guaranteed by construction that those four points will remain in the same place in the diagram once you update the Icon extent, and this entails that all other points will also remain in the same place, since it's a linear transfomation.

**Comment by casella** [comment]
> Maybe I'm missing something, but doing it that way seems like it would be unusably slow.

Please check the use case in https://github.com/PowerGrids/PowerGrids/issues/121. The alternative is to open the 70+ models in the library and manually resize those icons one by one. Now that will be unusably slow, and unnecessarily painful.

I don't really care about how long this takes, it could even be 30 minutes or two hours. You can have a break and relax or do something else in the meantime. The alternative to that is hours of really, really dumb manual work, or living forever with the consequences of a badly chosen initial extent, which in this case means that handling the diagrams is really awkward.

Of course if you see smarter ways to accomplish the same effect of my pseudo-code, that would be great. My goal now is to quickly get have an implementation that works and spares us the pain of the manual conversion. If it's not efficient, it doesn't matter, you only do it once. The important thing is that it be automatic. Of course it should be activated optionally, with a warning like "This may take several minutes or more".

> Also, I'm not sure I really understand the issue

Please check the use case in https://github.com/PowerGrids/PowerGrids/issues/121 and it should be immediately clear what the problem is.

> but to me it seems really odd that changing an annotation in a class would automatically change the annotation of components of that class.

For Icon extent annotations, that's the only sane thing to do if you don't want to screw up existing diagrams, see https://github.com/PowerGrids/PowerGrids/issues/121.

> That seems more like something that should be done with a conversion script, though I guess conversion scripts in Modelica does not support such changes at the moment.

In principle, yes. In this case we are releasing a new version which is not backwards compatible and with no option for conversion scripts, because there are not so many users around of the first version. However, we need to update all the example cases of the library that we are inheriting for the previous version.

The general use case is the following (which is exactly what happened to us): you are writing a new library and you start with some icons with some extents, and use them to build example and test cases. After a while you recognize that the extents you chose are not good, for some reason. At that point you would like to be able to change them, without screwing up the graphical appearance of all the examples and tests that you have already built.

It's not a conversion feature, it's more a refactoring feature, like changing the name of a class should change the name of all the declarations of instances of the same class in the loaded package. That's also something that comes in very handy when you build new libraries, as you may have second thoughts about naming stuff.

@perost is the use case clear now?

**Comment by perost** [comment]
So if I understand correctly, when changing the extents of a class' icon you want to scale the icon of any component of that class with the proportional change in the class' icon size? Implementing something like that properly will probably require a lot of effort, we don't have anything like that in the compiler right now.

**Comment by casella** [comment]
There is one alternative solution that is much cheaper and is entirely upon the GUI.

Currently, when you click on the extent of a component in a diagram, it gets selected, no matter whether there is actually any drawn item of the component under the pointer tip. This is what cause trouble in PowerGrids, because you have nearby components whose icons sit nicely side-by-side, but whose icon _extents_ overlap extensively. In this case you often end up fighting the GUI in the vane attempt of getting the right icon to be selected.

The alternative, which would make any library where the extents of the icons were not carefully crafted to match the actual extent of the icon, is to only select the component if the pointer tip is actually above some item. This is how Dymola works, and it works just fine with PowerGrids as we have it now.

@adeas31 this can be implemented very cheaply: you already have a method to figure out the smallest extent that includes all items in a diagram (this was asked by @ceraolo to resize the diagram's extent to its proper size). You can just re-use that to determine what is the "virtual extent" that contains the actual graphical items when opening the model, and then use that virtual extent (maybe with some 10% extra margin) instead of the actual extent to decide when the component is selected.

Seems easily implemented to me. Would that work for you?

**Comment by casella** [comment]
> This would determine the smallest bonding box (which is ok). Dymola does something different, is able to see whether the mouse pointer is on an object (text, shape, etc) Or an empty space

If the QT primitives allow getting that behaviour without programming it explicitly, I would go for it. Otherwise I guess a bounding box should be good enough.

**Comment by ceraolo** [comment]
Since the OM GUI has become very sluggish recently (probably in the near future I will supply a significant example) , I would choose the option that creates the minimum overhead on it.

**Comment by casella** [comment]
@adeas31 when you look into this, I would suggest you to try editing `PowerGrids.Examples.IEEE14bus.IEEE14busPowerFlow`. The user experience doing this once your changes are done should be smooth, and definitely not frustrating 😅

**Comment by adeas31** [comment]
> @adeas31 when you look into this, I would suggest you to try editing `PowerGrids.Examples.IEEE14bus.IEEE14busPowerFlow`, i.e., changing parameters and moving components around. The user experience doing this once your changes are done should be smooth, and definitely not frustrating 😅

Changing parameters requires new call to instance API so yes it will take some time. Moving components around should be quick, see my comment in #11887

**Comment by adeas31** [comment]
> > This would determine the smallest bonding box (which is ok). Dymola does something different, is able to see whether the mouse pointer is on an object (text, shape, etc) Or an empty space
> 
> If the QT primitives allow getting that behaviour without programming it explicitly, I would go for it. Otherwise I guess a bounding box should be good enough.

PR #11928 allows you to select the component by clicking on some drawn shape instead of the bounding rectangle of the component. You can test with tomorrow's nightly build, if you are satisfied with the solution then I will port this to 1.22

**Comment by casella** [comment]
Great! @ceraolo, the honour is yours :)

**Comment by casella** [comment]
@AndreaBartolini you can also check this with tomorrow's nightly build. If you can handle PowerGrids diagrams such as the IEEE14 bus examples without a fuss, then we can have @adeas31 port it to 1.22.2 and close this ticket for good. Thanks!

**Comment by adeas31** [comment]
@ceraolo and @AndreaBartolini any update on this?

**Comment by casella** [comment]
@AndreaBartolini was off in Stockholm for a couple of days after the OpenModelica workshop. I guess he can test this later this week.

**Comment by casella** [comment]
If I could download the nightly build I'd try it out myself, but I can't do it from LiU's network, for some reason...

**Comment by ceraolo** [comment]
Sorry for being late, but my Ubuntu OM did not update for a few days.
Today I could check 1.23.0-dev-297.
I think now OM works as expected. Thank you, @adeas31!

BTW, @AndreaBartolini, now I would recommend PowerGrids transformer icons to have their two circles being drawn over a white rectangle, otherwise it is difficult to grab them. But this is something outside this ticket.

**Comment by adeas31** [comment]
@ceraolo with PR #11968 the selection behavior is improved. Basically when a shape is part of an icon then I consider the whole shape for selection instead of just the outline of the shape. Dymola is pretty much doing the same. This will be problematic when you have some transparent shape of icon1 on top of icon2. With this change the icon1 will always be selected.

The real problem are the line shapes. For example, try selecting the ground icon. You have to very carefully click on the thin lines. For now, I have no idea how we can improve this.

Of course as a library developer you can make filled shape to overcome the current behavior.

**Comment by AndreaBartolini** [comment]
I'm not sure to have deeply understood the question, but I prefer to avoid the white background, if possible, in order to avoid to hide something.
When the last PR will be available (I suppose tomorrow) I'll check in detail, I think it's enough to increase the tickness of the lines in the icons.

**Comment by perost** [comment]
> The real problem are the line shapes. For example, try selecting the ground icon. You have to very carefully click on the thin lines. For now, I have no idea how we can improve this.

If nothing gets selected using the shapes, maybe fall back to using the bounding boxes? Maybe I'm not a typical user, but I would be surprised if clicking on an icon's bounding box didn't select it if there's no ambiguity on that I was trying to click on.

In fact, it might be better to check the bounding boxes first and only check the shapes if the cursor intersects with multiple boxes, since I assume checking arbitrary shapes can be very expensive.

**Comment by adeas31** [comment]
> > The real problem are the line shapes. For example, try selecting the ground icon. You have to very carefully click on the thin lines. For now, I have no idea how we can improve this.
> 
> If nothing gets selected using the shapes, maybe fall back to using the bounding boxes? Maybe I'm not a typical user, but I would be surprised if clicking on an icon's bounding box didn't select it if there's no ambiguity on that I was trying to click on.
> 
> In fact, it might be better to check the bounding boxes first and only check the shapes if the cursor intersects with multiple boxes, since I assume checking arbitrary shapes can be very expensive.

The default behavior was to check for bounding box. We changed this to overcome the issue reported in this ticket. If we fallback to bounding box then we will have the same problem as reported.

**Comment by perost** [comment]
> > > The real problem are the line shapes. For example, try selecting the ground icon. You have to very carefully click on the thin lines. For now, I have no idea how we can improve this.
> > 
> > 
> > If nothing gets selected using the shapes, maybe fall back to using the bounding boxes? Maybe I'm not a typical user, but I would be surprised if clicking on an icon's bounding box didn't select it if there's no ambiguity on that I was trying to click on.
> > In fact, it might be better to check the bounding boxes first and only check the shapes if the cursor intersects with multiple boxes, since I assume checking arbitrary shapes can be very expensive.
> 
> The default behavior was to check for bounding box. We changed this to overcome the issue reported in this ticket. If we fallback to bounding box then we will have the same problem as reported.

The issue as I understood it was that multiple icons can overlap in unexpected ways, making it hard to select the one you want. But my suggestion is to use the bounding boxes _unless_ the cursor intersects multiple boxes, in which case you can check the shapes to narrow down the selection. But if a user clicks on a bounding box that doesn't intersect with anything else then I don't see why we shouldn't select the icon.

**Comment by adeas31** [comment]
Sure. In the case of one icon we can always fallback to bounding box. But it will not solve the actual issue. We could have the case when lets say ground and resister bounding boxes overlap and you want to select the ground.

**Comment by perost** [comment]
> Sure. In the case of one icon we can always fallback to bounding box. But it will not solve the actual issue. We could have the case when lets say ground and resister bounding boxes overlap and you want to select the ground.

If the bounding boxes (not the extents) are overlapping, then the shapes of the icons are probably overlapping too since most icons are at least vaguely rectangular in my experience. In that situation the user probably just wants to move the icons so they don't overlap anymore, and it probably doesn't matter much if they are slightly hard to select.

An alternative if you really want to make sure something is selected would be to select based on which shape is closest to the cursor, but I think that just risks looking arbitrary to users if they're not aware of exactly what OMEdit is doing. 

Personally I would just expect OMEdit to always select based on the bounding boxes and the icons' depth, since I view an icon as one entity rather than a collection of shapes. If the bounding boxes are overlapping so much that it's an issue then being able to only select the top-most one is enough to disentangle them. That's how most graphical software works after all, but maybe modeling tools have different expectations.

But overlapping icons are the exception, and we shouldn't make it hard to select non-overlapping icons because of that.

**Comment by ceraolo** [comment]
This ticket was initiated by something I wrote in another ticket, so I feel I should contribute.

We have three solutions:
1. solution after #11968. this improves changes after #11928. I did not test it, but the description in https://github.com/OpenModelica/OpenModelica/issues/11866#issuecomment-1933972047 is pretty clear. It would make OM work nearly as Dymola does these days.  In reality Dymola is smarter with icons such as ground.
2. solution as proposed by @perost: "check the bounding boxes first and only check the shapes if the cursor intersects with multiple boxes" . 

We could evaluate these on two significant examples.
e1) _transformer over another icon_. For this case the solution in PR #11928 was not good, but AFAIU #11968  solves this case
e2) _ground over another icon_. For this case solution in PR #11968 is not enough, as @adeas31 says. This case is not an issue in Dymola, but we do not know how they solved it. 

So, if solution 2 does not solve e2, I would consider it equivalent to solution 1, which is my preferred, since is very near to Dymola's, and this eases users which work with both environments or want to switch from Dymola to OM.

If solution solves e2, probably it should be chosen. I am not sure that this happens, however. 

To conclude, for me both 1 and 2 are good enough, because example e2 is not very important, since it represents a small set of cases.

**Comment by casella** [comment]
> The default behavior was to check for bounding box. 

For the record, it was to check for the extent, which was the really problematic thing, because some of the extents were partially empty, so they could easily overlap when the icons were placed in the diagram.

**Comment by adeas31** [comment]
> > The default behavior was to check for bounding box.
> 
> For the record, it was to check for the extent, which was the really problematic thing, because some of the extents were partially empty, so they could easily overlap when the icons were placed in the diagram.

It is the same thing. The bounding box is created using the extent. For example,

```Modelica
model A
annotation(
    Icon(graphics = {Rectangle(extent = {{-60, 60}, {60, -60}})}, coordinateSystem(extent = {{-100, -100}, {100, 100}})));
end A;

model B
  A a annotation(
    Placement(transformation(origin = {-30, 30}, extent = {{-10, -10}, {10, 10}})));
end B;
```

The icon of `A` in `B` has the bounding box that is created using the extent defined in the coordinate system of `A` i.e., `coordinateSystem(extent = {{-100, -100}, {100, 100}})`. Then we apply the transformation (scaling, flipping etc.) on it using the extent defined in the placement annotation of component `a` defined in `B` i.e., `extent = {{-10, -10}, {10, 10}}`.

**Comment by perost** [comment]
> > > The default behavior was to check for bounding box.
> > 
> > 
> > For the record, it was to check for the extent, which was the really problematic thing, because some of the extents were partially empty, so they could easily overlap when the icons were placed in the diagram.
> 
> It is the same thing. The bounding box is created using the extent. For example,
> 
> ```modelica
> model A
> annotation(
>     Icon(graphics = {Rectangle(extent = {{-60, 60}, {60, -60}})}, coordinateSystem(extent = {{-100, -100}, {100, 100}})));
> end A;
> 
> model B
>   A a annotation(
>     Placement(transformation(origin = {-30, 30}, extent = {{-10, -10}, {10, 10}})));
> end B;
> ```
> 
> The icon of `A` in `B` has the bounding box that is created using the extent defined in the coordinate system of `A` i.e., `coordinateSystem(extent = {{-100, -100}, {100, 100}})`. Then we apply the transformation (scaling, flipping etc.) on it using the extent defined in the placement annotation of component `a` defined in `B` i.e., `extent = {{-10, -10}, {10, 10}}`.

What I meant was the smallest possible rectangle that contains the shapes of the icon in the currently shown diagram.

**Comment by ceraolo** [comment]
A new observation:
moving or opening the parameters dialog box for ground (and inductance and the like) can be done clicking on the text boxes, the instance name or a numerical value, when exists. 
No need to click on the tiny lines.

**Comment by casella** [comment]
> What I meant was the smallest possible rectangle that contains the shapes of the icon in the currently shown diagram.

Yes, this was my original proposal. Use the same functionality that is used to recompute the smallest possible extent on diagrams using this button  ![immagine](https://github.com/OpenModelica/OpenModelica/assets/4233148/58947061-022b-45c7-986e-17c37ce8d33d)

**Comment by adeas31** [comment]
PR #11978 selects the item based on the minimum bounding box. The minimum bounding box is calculated by doing a bitwise OR of the bounding boxes of all the shapes of the item.

The gives improved selection behaviour for PowerGrids models. The selection of items like ground, fixed, spring etc. are also very easy with this implementation.

**Comment by casella** [comment]
Great! I'll test this ASAP on a hand-triggered nightly build and report

**Comment by adeas31** [comment]
You can now start the nightly build and test this.

**Comment by adeas31** [comment]
@casella the changes are ported to the maintenance branch. If you are satisfied with the solution then I guess we can make 1.22.2 and move the rest of the issues to 1.22.3 or maybe 1.23

**Comment by casella** [comment]
We just tested this feature in the new nightly build, it works like a charm. We can proceed with 1.22.2.
